transformer_engine/pytorch/module/layernorm_linear.py

```diff
@@ -894,6 +901,10 @@ def wgrad_gemm(
                     # Deallocate input tensor if permitted
                     if not ctx.return_layernorm_output:
                         clear_tensor_data(ln_out_total)
+                    elif (
+                        (ln_out_total is not ln_out) or not ctx.ln_out_return_is_ln_out
+                    ) and not ctx.ub_overlap_ag:
+                        clear_tensor_data(ln_out_total)
```

Comment:
Two issues:
- In the backward pass, `ub_overlap_ag` is for the grad output all-gather. We should use `ub_bulk_dgrad` for the input all-gather.
- Isn't `ln_out_return_is_ln_out` redundant? We should double check, but I think the correct logic is:
```suggestion
                    if not ctx.return_layernorm_output and not ctx.return_layernorm_output_gathered:
                        clear_tensor_data(ln_out)
                    elif ctx.return_layernorm_output_gathered and ctx.ln_out_needs_gather:
                        clear_tensor_data(ln_out)
                    if ctx.ln_out_needs_gather and not ctx.ub_bulk_dgrad:
                        clear_tensor_data(ln_out_total)
```